File: 8. Count leaf nodes.cpp
#/*
 Task: Count leaf nodes
 Time Complexity: O(n)
 Space Complexity: O(h) — recursion stack
*/
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
int countLeaves(Node* root)
{
    if (!root)
        return 0;
    if (!root->left && !root->right)
        return 1;
    return countLeaves(root->left) + countLeaves(root->right);
}
int main()
{
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    cout << "Leaves: " << countLeaves(root) << endl;
    return 0;
}
// sc tc

File: 8. Count leaf nodes.py
"""
Task: Count leaf nodes in binary tree
Time Complexity: O(n)
Space Complexity: O(h)
"""

class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def count_leaf(root):
    if not root: return 0
    if not root.left and not root.right: return 1
    return count_leaf(root.left)+count_leaf(root.right)

if __name__=='__main__':
    root=Node(1); root.left=Node(2); root.right=Node(3); root.left.left=Node(4)
    print('Leaf count:', count_leaf(root))
